#8773 - On restart, show the search history tab, if search history exists

varad64 opened the pull request:
Fix to address #8773:

Calculate search recents list size correctly to set the correct tab in the SearchFragment's layout depending on whether or not, a search history exists.

On app restart, if a search history exists, the search history tab is shown instead of the categories tab, to achieve expected behavior.

Thanks!

biodranik commented:
Thanks, but the behavior in this PR does not match what was described in the issue.

In short:
- The last used tab should be restored even after the app restart (the current behavior without restart is ok, the same behavior should be with the app restart).
- If the History tab is still empty (or if history is disabled in settings), there's no need to restore it. Categories should be displayed instead.

varad64 commented:
Ohh alright, my bad!
Thanks for the review, let me rework this and provide an update.
Thank you!

varad64 commented:
Hello @biodranik, I have updated the code now and made use of SharedPreferences to remember the last tab selected and restore it on app restart.

Could you please review the changes when possible?

Thanks in advance!

biodranik commented on android/app/src/main/java/app/organicmaps/search/TabAdapter.java:
```diff
@@ -11,6 +12,8 @@
 
 import com.google.android.material.tabs.TabLayout;
+
+import app.organicmaps.MwmApplication;
```
Is it strictly necessary?

varad64 replied:
Yes @biodranik, to access the static method `prefs`

biodranik commented on android/app/src/main/java/app/organicmaps/search/TabAdapter.java:
```diff
@@ -72,8 +75,9 @@ private class PageChangedListener extends TabLayout.TabLayoutOnPageChangeListene
     {
       super.onPageSelected(position);
-      if (mTabSelectedListener != null)
+      if (mTabSelectedListener != null) {
         mTabSelectedListener.onTabSelected(Tab.values()[position]);
+      }
```
Please follow the repo's code style.
```suggestion
      if (mTabSelectedListener != null)
        mTabSelectedListener.onTabSelected(Tab.values()[position]);
```

varad64 replied:
I will make the necessary changes and provide an update, thanks!

biodranik commented on android/app/src/main/java/app/organicmaps/search/TabAdapter.java:
```diff
@@ -91,6 +95,9 @@ private static class OnTabSelectedListenerForViewPager extends TabLayout.ViewPag
     public void onTabSelected(TabLayout.Tab tab)
     {
+      SharedPreferences.Editor editor = MwmApplication.prefs(mContext).edit();
+      editor.putInt("lastSelectedTab", tab.getPosition());
```
Unique constants should be used as keys to avoid copy-paste and mistakes.

varad64 replied:
Okay, I will make the change, thanks!

biodranik reviewed (commented):
Thanks! Did you test the behavior in different scenarios?

What is the behavior when search history is disabled?

varad64 commented:
> Thanks! Did you test the behavior in different scenarios?
> 
> What is the behavior when search history is disabled?

Yes @biodranik.

I had actually not tested the behvior when the history is disabled. I just tested it out now and hen the search history is disabled, only the Categories tab is shown and works as expected.

When the search history is enabled and the user initiates search for the first time, the Categories tab is shown. Then, depending on the last tab selected before exit, the same tab is restored the next time the search fragment is shown.

Thank you!

biodranik commented:
> I had actually not tested the behvior when the history is disabled. I just tested it out now and hen the search history is disabled, only the Categories tab is shown and works as expected.

Did you test this scenario:
1. Use/save search history tab
2. Disable search history
3. Open search again (categories should open, even if history tab was saved as the last used one)

?

varad64 commented:
> > I had actually not tested the behvior when the history is disabled. I just tested it out now and hen the search history is disabled, only the Categories tab is shown and works as expected.
> 
> Did you test this scenario:
> 
> 1. Use/save search history tab
> 2. Disable search history
> 3. Open search again (categories should open, even if history tab was saved as the last used one)
> 
> ?

Hello @biodranik,

Yes, this is the observed behavior. Thanks!

biodranik commented on android/app/src/main/res/values/strings.xml:
```diff
@@ -2263,4 +2263,5 @@
     <string name="type.amenity.dojo">Dojo</string>
     <!-- https://wiki.openstreetmap.org/wiki/Tag:leisure=sports_hall -->
     <string name="type.leisure.sports_hall">Sports hall</string>
+    <string name="last_selected_tab">lastSelectedTab</string>
```
Please add a constant into android/app/src/main/res/values/donottranslate.xml, this file is automatically regenerated by scripts.

biodranik reviewed (commented):
Thanks! LGTM after fixing the comment.

CC @pastk

pastk reviewed (commented):
The code LGTM.

Please rebase and squash your commits into a single one (and there should be no merge commits).

varad64 commented:
Thanks @pastk and @biodranik for your comments.
And apologies it took me a while to a while to take the necessary action.
I have performed the needful steps, rebased commits into a single commit and the changes are now ready for review.

Thank you!

pastk commented:
> I have performed the needful steps, rebased commits into a single commit and the changes are now ready for review.

There are still 6 commits in this PR :)

varad64 commented:
Yeah, sorry about the confusion!
I have rebased and squashed on my local branch properly, however I must have messed something up while pushing it.
Let me check what I've done wrong and rectify the problem.

Thanks!

pastk commented:
You need to force-push (e.g. `git push --force origin HEAD`) to overwrite the commit history (remove the un-squashed commits and replace with a new single commit).

varad64 commented:
Hello @pastk, I appreciate you trying to help me.
But I was unable to get the command to work as intended, I reckon.
My sincere apologies!

I will look into any alternatives, if available to overwrite the commit history cleanly.
Thank you!

varad64 commented:
Thank you very much, @pastk
I will rectify and rebase properly.

varad64 commented:
Hello @pastk, it took me a while to get things right, but finally was able to rebase and merge correctly.
Thank you very much for help with the docs!

varad64 commented on android/app/src/main/java/app/organicmaps/search/TabAdapter.java:
```diff
@@ -91,6 +94,9 @@ private static class OnTabSelectedListenerForViewPager extends TabLayout.ViewPag
     public void onTabSelected(TabLayout.Tab tab)
     {
+      SharedPreferences.Editor editor = MwmApplication.prefs(mContext).edit();
+      editor.putInt(mContext.getString(R.string.last_selected_tab), tab.getPosition());
```
Thanks for the review, @pastk.

I have updated the code to use a string constant declared in `Config.java` and accordingly replaced all usages of the string constant.

However, I am unable to test my changes due to the build failing because of a `CMake` error within the `cmake/OmniHelpers.cmake` file's `add_library` function.

Initially, I thought the build was breaking as I recently updated Android Studio to LadyBug. However, that does not seem to be the issue.

I can verify that NDK version 27.2.12479018 is installed and so is cmake 3.22.1.

I also tried reinstalling both the NDK and cmake to ensure just in case they could have been broken during the Android Studio update.

The error still persists and I am unable to figure out what is causing the issue.

Here's the relevant errors from the log, I have also attached the complete log file for reference:
```
CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "coding" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  coding/CMakeLists.txt:94 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "descriptions" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  descriptions/CMakeLists.txt:10 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "drape" links to target "expat::expat" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?
Call Stack (most recent call first):
  drape/CMakeLists.txt:164 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "drape" links to target "expat::expat" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?
Call Stack (most recent call first):
  drape/CMakeLists.txt:164 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "drape_frontend" links to target "expat::expat" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
Call Stack (most recent call first):
  drape_frontend/CMakeLists.txt:211 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "drape_frontend" links to target "expat::expat" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
Call Stack (most recent call first):
  drape_frontend/CMakeLists.txt:211 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "editor" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  editor/CMakeLists.txt:38 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "ge0" links to target "expat::expat" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?
Call Stack (most recent call first):
  ge0/CMakeLists.txt:12 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "mwm_diff" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  generator/mwm_diff/CMakeLists.txt:8 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "indexer" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  indexer/CMakeLists.txt:156 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "kml" links to target "expat::expat" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?
Call Stack (most recent call first):
  kml/CMakeLists.txt:29 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "map" links to target "expat::expat" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?
Call Stack (most recent call first):
  map/CMakeLists.txt:87 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "map" links to target "expat::expat" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?
Call Stack (most recent call first):
  map/CMakeLists.txt:87 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "platform" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  platform/CMakeLists.txt:183 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "routing" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  routing/CMakeLists.txt:187 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "routing_common" links to target "expat::expat" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
Call Stack (most recent call first):
  routing_common/CMakeLists.txt:18 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "search" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  search/CMakeLists.txt:173 (omim_add_library)


CMake Error at shaders/CMakeLists.txt:3 (add_library):
  Target "shaders" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at shaders/CMakeLists.txt:3 (add_library):
  Target "shaders" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "storage" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  storage/CMakeLists.txt:64 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "tracking" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  tracking/CMakeLists.txt:21 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "traffic" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  traffic/CMakeLists.txt:12 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "transit" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  transit/CMakeLists.txt:21 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "organicmaps" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  android/app/src/main/cpp/CMakeLists.txt:73 (omim_add_library)


CMake Error at cmake/OmimHelpers.cmake:55 (add_library):
  Target "organicmaps" links to target "expat::expat" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  android/app/src/main/cpp/CMakeLists.txt:73 (omim_add_library)


CMake Warning:
  Manually-specified variables were not used by the project:

    OS


CMake Generate step failed.  Build files cannot be regenerated correctly.
	at com.android.build.gradle.internal.cxx.process.ExecuteProcessKt.execute(ExecuteProcess.kt:288)
	at com.android.build.gradle.internal.cxx.process.ExecuteProcessKt$executeProcess$1.invoke(ExecuteProcess.kt:108)
	at com.android.build.gradle.internal.cxx.process.ExecuteProcessKt$executeProcess$1.invoke(ExecuteProcess.kt:106)
	
	
Caused by: com.android.ide.common.process.ProcessException: Error while executing process /Users/varad/Library/Android/Sdk/cmake/3.22.1/bin/cmake with arguments {-H/Users/varad/Code/android-apps/organicmaps -DCMAKE_SYSTEM_NAME=Android -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_SYSTEM_VERSION=21 -DANDROID_PLATFORM=android-21 -DANDROID_ABI=arm64-v8a -DCMAKE_ANDROID_ARCH_ABI=arm64-v8a -DANDROID_NDK=/Users/varad/Library/Android/Sdk/ndk/27.2.12479018 -DCMAKE_ANDROID_NDK=/Users/varad/Library/Android/Sdk/ndk/27.2.12479018 -DCMAKE_TOOLCHAIN_FILE=/Users/varad/Library/Android/Sdk/ndk/27.2.12479018/build/cmake/android.toolchain.cmake -DCMAKE_MAKE_PROGRAM=/Users/varad/Library/Android/Sdk/cmake/3.22.1/bin/ninja -DCMAKE_C_FLAGS=-fno-function-sections -fno-data-sections -Wno-extern-c-compat -DCMAKE_CXX_FLAGS=-fexceptions -frtti -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/Users/varad/Code/android-apps/organicmaps/android/app/build/intermediates/cxx/Debug/12524z14/obj/arm64-v8a -DCMAKE_RUNTIME_OUTPUT_DIRECTORY=/Users/varad/Code/android-apps/organicmaps/android/app/build/intermediates/cxx/Debug/12524z14/obj/arm64-v8a -DCMAKE_BUILD_TYPE=Debug -B/Users/varad/Code/android-apps/organicmaps/android/app/nativeOutputs/Debug/12524z14/arm64-v8a -GNinja -DANDROID_TOOLCHAIN=clang -DANDROID_STL=c++_static -DOS=mac os x -DSKIP_TESTS=ON -DSKIP_TOOLS=ON -DUSE_PCH=OFF -DENABLE_VULKAN_DIAGNOSTICS=OFF -DENABLE_TRACE=OFF}
	at com.android.build.gradle.internal.process.GradleProcessResult.buildProcessException(GradleProcessResult.java:73)
	at com.android.build.gradle.internal.process.GradleProcessResult.assertNormalExitValue(GradleProcessResult.java:48)
	at com.android.build.gradle.internal.cxx.process.ExecuteProcessKt.execute(ExecuteProcess.kt:277)
	... 183 more
Caused by: org.gradle.process.internal.ExecException: Process 'command '/Users/varad/Library/Android/Sdk/cmake/3.22.1/bin/cmake'' finished with non-zero exit value 1
	at org.gradle.process.internal.DefaultExecHandle$ExecResultImpl.assertNormalExitValue(DefaultExecHandle.java:442)
	at com.android.build.gradle.internal.process.GradleProcessResult.assertNormalExitValue(GradleProcessResult.java:46)
	... 184 more
```

Seeing as the app was building and working correctly a month ago, I think the update could have broken something but am not sure and would like to confirm if something else could be the problem.

Thank you in advance!
[build_error.log](https://github.com/user-attachments/files/18939071/build_error.log)

pastk replied:
all those errors mention `expat` which is a 3party c library
it should be in `3party/expat/expat/`

please check that all repo's submodules are cloned, e.g. `git submodule update --init --recursive --depth=1`

varad64 replied:
Hey @pastk, thanks for taking a look.

I found that `expat` is present, within `3party/expat/` as expected.
Should I try and re-clone the repo's submodules, just in case the symbolic link is broken?

pastk replied:
also try to reset caches in android studio

varad64 replied:
Sorry if I was not clear, they are present at the expected location. PFA a screenshot showing the files contained in the `expat` directory:
<img width="1391" alt="Screenshot 2025-02-24 at 7 11 29 PM" src="https://github.com/user-attachments/assets/62cfc50b-2a78-4780-9d6e-e275033c018e" />

I have reset the cache in android studio, also cleaned and rebuilt the project multiple times. Things are still the same!

varad64 replied:
No @pastk, I hadn't.

I just updated AGP to 8.8.1 and still see the errors.
My branch is master from 2 days ago.

pastk replied:
I'm out of ideas then... Maybe ask in the TG chat https://t.me/OrganicMaps

pastk replied:
Maybe checkout a version from a couple months ago to see if it was some recent changes in OM sources..

varad64 replied:
Hey @pastk, the app is now building and running correctly.

Your advice
> try to git submodule deinit -f 3party/expat and then git submodule update --init --recursive

seemed to resolve the issue.

Thank you very much for your time and advice!

All tests still passing, after using the string constant (declared in `Config.java`) instead of the string resource.

I will rebase and pass it over for review in some time.

pastk commented on android/app/src/main/java/app/organicmaps/util/Config.java:
```diff
@@ -36,6 +36,8 @@ public final class Config
   private static final String KEY_DONATE_URL = "DonateUrl";
   private static final String KEY_PREF_SEARCH_HISTORY = "SearchHistoryEnabled";
 
+  public static final String KEY_PREF_LAST_SELECTED_TAB = "LastSelectedTab";
```
`LastSearchTab` is more obvious

varad64 replied:
Thanks for the suggestion, @pastk.
Assignment and usages updated to `LastSearchTab` and `KEY_PREF_LAST_SEARCHED_TAB` respectively.

pastk reviewed (commented):
Good, but some minor things left.

varad64 commented on android/app/src/main/res/values/donottranslate.xml:
Code updated to match the source.

pastk reviewed (approved):
Thanks!
